packages/utils/src/internals/robots.ts

```diff
@@ -0,0 +1,126 @@
+        } catch (e) {
+            if (e instanceof HTTPError && e.response.statusCode === 404) {
+                return new RobotsFile({ isAllowed() { return true; }, getSitemaps() { return []; } }, proxyUrl);
+            }
+            throw e;
+        }
+    }
+
+    isAllowed(url: string): boolean {
+        return this.robots.isAllowed(url, '*') ?? false;
+    }
+
+    getSitemaps(): string[] {
+        return this.robots.getSitemaps();
+    }
+
+    async parseSitemaps(): Promise<Sitemap[]> {
+        return Promise.all(this.robots.getSitemaps().map(async (sitemap) => Sitemap.load(sitemap, this.proxyUrl)));
+    }
+
+    async parseUrlsFromSitemaps(): Promise<string[]> {
+        return (await this.parseSitemaps()).flatMap((sitemap) => sitemap.urls);
+    }
+}
+
+export class Sitemap {
+    constructor(readonly urls: string[]) {}
+
+    static async load(url: string, proxyUrl?: string): Promise<Sitemap> {
+        const parsingState: {
+            context: 'sitemapindex' | 'urlset' | undefined;
+            loc: boolean;
+            urls: string[];
+            sitemapUrls: string[];
+            visitedSitemapUrls: string[];
+        } = { context: undefined, loc: false, urls: [], sitemapUrls: [url], visitedSitemapUrls: [] };
+
+        while (parsingState.sitemapUrls.length > 0) {
+            const sitemapUrl = parsingState.sitemapUrls.pop()!;
+            parsingState.visitedSitemapUrls.push(sitemapUrl);
+
+            const response = await gotScraping({ url: sitemapUrl, proxyUrl, responseType: 'text', method: 'GET' });
+            const parser = sax.parser(true);
+
+            parser.onopentag = (node) => {
+                if (node.name === 'loc' && parsingState.context !== undefined) {
+                    parsingState.loc = true;
+                }
+                if (node.name === 'urlset') {
+                    parsingState.context = 'urlset';
+                }
+                if (node.name === 'sitemapindex') {
+                    parsingState.context = 'sitemapindex';
+                }
+            };
+
+            parser.onclosetag = (name) => {
+                if (name === 'loc') {
+                    parsingState.loc = false;
+                }
+            };
+
+            parser.ontext = (text) => {
+                if (parsingState.loc) {
+                    if (parsingState.context === 'sitemapindex') {
+                        if (!parsingState.visitedSitemapUrls.includes(text)) {
+                            parsingState.sitemapUrls.push(text);
+                        }
+                    }
+                    if (parsingState.context === 'urlset') {
+                        parsingState.urls.push(text);
+                    }
+                }
+            };
+
+            if (response.statusCode === 200) {
+                try {
+                    parser.write(response.body).close();
+                } catch (e) {
```

Comment:
If we're not using the error, we don't need the error binding!

```suggestion
                } catch {
```

Reply:
Cool, it haven't occurred to me that you can do that :slightly_smiling_face: